pages/login/LoginHandler.inc.php

```diff
@@ -213,57 +220,58 @@ function resetPassword($args, $request) {
-			$newPassword = Validation::generatePassword();
+		$passwordResetForm = new ResetPasswordForm($user, $request->getSite(), $confirmHash);
+		$passwordResetForm->initData();
 
-			if ($user->getAuthId()) {
-				$authDao = DAORegistry::getDAO('AuthSourceDAO'); /* @var $authDao AuthSourceDAO */
-				$auth = $authDao->getPlugin($user->getAuthId());
-			}
+		
+		$passwordResetForm->validatePasswordResetHash($request)
+			? $passwordResetForm->display($request)
+			: $passwordResetForm->displayInvalidHashErrorMessage($request);
+	}
 
-			if (isset($auth)) {
-				$auth->doSetUserPassword($user->getUsername(), $newPassword);
-				$user->setPassword(Validation::encryptCredentials($user->getId(), Validation::generatePassword())); // Used for PW reset hash only
-			} else {
-				$user->setPassword(Validation::encryptCredentials($user->getUsername(), $newPassword));
-			}
+	/**
+	 * Reset a user's password
+	 * @param $args array first param contains the username of the user whose password is to be reset
+	 */
+	public function updateResetPassword($args, $request)
+	{
+		$this->_isBackendPage = true;
+		$this->setupTemplate($request);
+		$templateMgr = TemplateManager::getManager($request);
 
-			$user->setMustChangePassword(1);
-			$userDao->updateObject($user);
+		$username = $request->getUserVar('username');
+		$userDao = DAORegistry::getDAO('UserDAO'); /** @var UserDAO $userDao */
+		$confirmHash = $request->getUserVar('hash');
 
-			// Send email with new password
-			$site = $request->getSite();
-			import('lib.pkp.classes.mail.MailTemplate');
-			$mail = new MailTemplate('PASSWORD_RESET');
-			$this->_setMailFrom($request, $mail, $site);
-			$mail->assignParams([
-				'username' => htmlspecialchars($user->getUsername()),
-				'password' => htmlspecialchars($newPassword),
-				'siteTitle' => htmlspecialchars($site->getLocalizedTitle()),
-			]);
-			$mail->addRecipient($user->getEmail(), $user->getFullName());
-			if (!$mail->send()) {
-				import('classes.notification.NotificationManager');
-				$notificationMgr = new NotificationManager();
-				$notificationMgr->createTrivialNotification($user->getId(), NOTIFICATION_TYPE_ERROR, array('contents' => __('email.compose.error')));
-			}
+		if ($username == null || ($user = $userDao->getByUsername($username)) == null) {
+			$request->redirect(null, null, 'lostPassword');
+		}
 
-			$templateMgr->assign(array(
-				'pageTitle' => 'user.login.resetPassword',
-				'message' => 'user.login.lostPassword.passwordSent',
-				'backLink' => $request->url(null, $request->getRequestedPage()),
-				'backLinkLabel' => 'user.login',
-			));
-			$templateMgr->display('frontend/pages/message.tpl');
+		import('lib.pkp.classes.user.form.ResetPasswordForm');
+
+		$passwordResetForm = new ResetPasswordForm($user, $request->getSite(), $confirmHash);
+		$passwordResetForm->readInputData();
+
+		if ( !$passwordResetForm->validatePasswordResetHash($request) ) {
+			$passwordResetForm->displayInvalidHashErrorMessage($request);
```

Comment:
I think this will continue to execute the code below, when it should prevent that from happening with e.g. a `return`.

Reply:
My mistake, yes it should `return` at that point . fixed it .